cca.go

```diff
@@ -0,0 +1,122 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import (
+	"errors"
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix"
+	"github.com/gonum/matrix/mat64"
+)
+
+// CanonicalCorrelations returns the `canonical' correlations between
+// the corresponding pairs of canonical variables -- described by the
+// left and right eigenvectors of the canonical correlation matrix,
+// pvecs and qvecs respectively. The canonical correlation matrix is
+// calculated using the singular value decompositions of the centered
+// inputs, and these singular value decompositions, xsvd and ysvd, are
+// also returned.
+//
+// The slice weights is used to weight the observations. If weights is nil,
+// each weight is considered to have a value of one, otherwise the length of
+// weights must match the number of observations or CanonicalCorrelations will
+// panic.
+//
+// Canonical Correlations will panic if the inputs x and y do not have the same
+// number of rows, and their columns should be interepretable as two different
+// sets of measurements (variables) on the same observations (each row representing
+// an observation).
+//
+// The sphered data are Xc * Sx^(-1/2) and Yc * Sy^(-1/2) respectively, and the
+// correlation matrix between the sphered data is the canonical correlation matrix,
+// Sx^(-1/2) * Sxy * Sy^(-1/2). In cases where S^(-1/2) is ambiguous, S^(-1/2) is
+// taken to be E * D^(-1/2) * E' where S can be eigendecomposed as S = E * D * E'.
+// Xc and Yc denote the centered input data matrices x and y (column means subtracted
+// from each column), Sx and Sy denote the sample covariance matrices within x and y
+// respectively, and Sxy denotes the covariance matrix between x and y.
+//
+// The first pair of left and right eigenvectors of the canonical correlation matrix
+// can be interpreted as directions into which the respective sphered data can be
+// projected such that the correlation between the two projections is maximised.
+// The second pair and onwards solve the same optimisation but under the constraint
+// that they are uncorrelated to previous projections.
+//
+// More details can be found at https://en.wikipedia.org/wiki/Canonical_correlation
+// or in Chapter 3 of
+// Koch, Inge. Analysis of multivariate and high-dimensional data.
+// Vol. 32. Cambridge University Press, 2013. ISBN: 9780521887939
+//
+// If one of the necessary singular value decompositions is not possible, an error
+// is returned as err and all other outputs are returned as nil.
+func CanonicalCorrelations(x, y mat64.Matrix, weights []float64) (corrs []float64, pvecs, qvecs *mat64.Dense, xsvd, ysvd *mat64.SVD, err error) {
+	n, _ := x.Dims()
+	yn, _ := y.Dims()
+	if n != yn {
+		panic("stat: unequal number of observations")
+	}
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	var ok bool
```

Comment:
I think this is needed because xsvd, ok = svdFactorizeCentered(... 
does not use `:=` this is because xsvd is already initialised because its defined as the name of the output variable to be returned. Should I rename it in order to remove this?